llumnix/llm_engine_manager.py

```diff
@@ -250,28 +256,110 @@ async def _migrate(self) -> None:
+                    dead_instances.add(instance_name)
+
+            if len(dead_instances) > 0:
+                self.scale_down(dead_instances, rebuild_migrate_backend=False)
+                # clear gloo migrate backend intermediate state
+                ray.kill(ray.get_actor("gloo_queue", "llumnix"))
+
+            return dead_instances
+
+        alive_instances = sorted(self.instances.keys())
+        pending_task = self.pending_rebuild_migration_instances
+        group_name = None
+
+        while len(alive_instances) > 0 and self.pending_rebuild_migration_instances > 0:
+            dead_instances = set()
+            group_name = random_uuid()
+            instance_rank = {instance_id: index for index, instance_id in enumerate(alive_instances)}
+            dead_instances = dead_instances.union(await run_task(alive_instances, "rebuild_migration_backend",
+                                                                  instance_rank, group_name))
+
+            if len(dead_instances) == 0 and self.pending_rebuild_migration_instances == pending_task:
+                dead_instances = dead_instances.union(await run_task(alive_instances, "warmup"))
+
+            if len(dead_instances) == 0:
+                self.pending_rebuild_migration_instances -= pending_task
+
+            alive_instances = sorted(set(self.instances.keys()) - dead_instances)
+            pending_task = self.pending_rebuild_migration_instances
+
+        if len(alive_instances) == 0:
+            self.pending_rebuild_migration_instances = 0
+            group_name = None
+
+        logger.info("rebuild {} migrate backend done, group_name: {}, alive instance ({}): {}"
+            .format(self.engine_manager_args.migration_backend, group_name, len(alive_instances), alive_instances))
+
+        # Restore migrate config
+        self.enable_migration = origin_config
+
     def scale_up(self, instance_id: Union[str, Iterable[str]], llumlet_actor_handles: List["ray.actor.ActorHandle"]) -> None:
         if isinstance(instance_id, str):
             instance_id = [instance_id,]
         instance_ids = list(instance_id)
+
+        indeed_update = False
+        no_pending_instance = (self.pending_rebuild_migration_instances == 0)
+
         for idx, ins_id in enumerate(instance_ids):
             if ins_id not in self.instances:
+                indeed_update = True
                 self.instances[ins_id] = llumlet_actor_handles[idx]
                 self.instance_migrating[ins_id] = False
+                self.pending_rebuild_migration_instances += 1
         self.global_scheduler.scale_up(instance_ids)
         self.num_instance = len(self.instances)
 
-    def scale_down(self, instance_id: Union[str, Iterable[str]]) -> None:
+        # When scaling up, we need to rebuild the migration backend. But if initially self.pending_rebuild_migration_instances != 0,
+        # a coroutine is already handling the membership change and it will account for the membership changes
```

Comment:
The word membership is confusing, why not simpliy called rebuild?

Reply:
get it
I will make it clearer